### PR TITLE
albef vqa finetune

### PR DESCRIPTION
**Pull request by katrina433**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #229
* #228
* #227
* #226
* #225
* __->__ #224

Differential Revision: [D38881553](https://our.internmc.facebook.com/intern/diff/D38881553)

### CODE REVIEW

**Comment by katrina433** [comment]
Refer to pr #174 for previous reviews, but this one will be landed as part of the stack.

File: torchmultimodal/utils/attention.py [hunk 1]
```diff
@@ -43,6 +45,32 @@ def get_extended_attention_mask(attention_mask: Tensor) -> Tensor:
     return extended_attention_mask
 
 
+def get_extended_attention_mask_for_decoder(attention_mask: Tensor) -> Tensor:
```
Comment (ebsmothers) [inline torchmultimodal/utils/attention.py:48]: I'm still not 100% sure why we need this. Why not just call get_causal_attention_mask and get_extended_attention_mask(attn_mask) then take the product in the decoder itself?

File: examples/albef/model.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        super().__init__()
+        self.model = model
+        self.answer_decoder = answer_decoder
+        self.loss = loss
+        self.answer_decoder_m = copy.deepcopy(self.answer_decoder)
+        remove_grad(
+            self.answer_decoder_m
+        )  # remove gradient for the momentum decoder model
+
+    def _train_forward(
```
Comment (ebsmothers) [inline examples/albef/model.py:166]: Both these forwards are hard to follow (I think this is unavoidable). But maybe a brief docstring for each method explaining at a high level the steps it is taking would help

**Review by ebsmothers — APPROVED** [review]
Looks good! Just a few more comments, but accepting to unblock you

File: examples/albef/configs/vqa.yaml [hunk 3]
```diff
@@ -0,0 +1,71 @@
+hidden_size: &hidden_size 768
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply examples/albef/configs/vqa.yaml:1]: Yeah the `_target_` approach is specific to Hydra. It could be a nice extension but for now I personally would not consider it mandatory. I think `&` is just used to mark an alias that can be reused later (i.e. if we want to use the same value for multiple configurables to constrain their values to match)

File: examples/albef/finetune.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+    dist_utils.init_distributed_mode(args)
+    device = torch.device(args.device)
+
+    seed = config["seed"] + dist_utils.get_rank()
+    torch.manual_seed(seed)
+    random.seed(seed)
+    cudnn.benchmark = True
+
+    datamodule = VQADataModule(**config["datamodule_args"])
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply examples/albef/finetune.py:180]: The alternative would just be to use data loader directly. I think we are trending in that direction (already had some conversations with Ankita about doing this for MDETR, which is in a similar state)

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38881553).

File: torchmultimodal/utils/attention.py [hunk 5]
```diff
@@ -27,14 +27,16 @@ def get_extended_attention_mask(attention_mask: Tensor) -> Tensor:
         extended_attention_mask (Tensor):
             The broadcastable attention mask, with the same dtype as ``attention_mask.dtype``.
     """
-    # We can provide a self-attention mask of dimensions [batch_size, query_seq_length, key_seq_length]
-    # ourselves in which case we just need to make it broadcastable to all heads,
-    # [batch_size, num_heads, query_seq_length, key_seq_length].
-    if attention_mask.dim() == 3:
+    if attention_mask.dim() == 4:
```
Comment (langong347) [inline torchmultimodal/utils/attention.py:30]: Can we add comments to this codepath as well? 
Also update the docstring since we are now accepting masks with head dim.

```
Accepting two types of attention masks:
        - Causal: masks that prevent attending to future positions of dimensions
            ``(batch_size, query_seq_len, key_seq_len)``
        - Padding: masks that prevent attending to token paddings of dimensions
            ``(batch_size, seq_len)``
```

Comment (langong347) [reply torchmultimodal/utils/attention.py:30]: Non blocking Q: this utility is ambiguous as it puts strong assumption around the original dims of the masks.
For example, dim=3 could mean both `(b, h, seq_len)` and `(b, q_seq_len, k_seq_len)`.
May introduce bugs unless we have a convention for the original dims. Better be careful with it.

Comment (ebsmothers) [reply torchmultimodal/utils/attention.py:30]: Yeah this is a fair point. I think we should not expect users to pass `(b, h, seq_len)` very often, since the inclusion of `h` implies they have already broadcasted (and if using this util the broadcasting will have already been to 4D). But agree there can be some ambiguity here. We could add a flag, but not sure that is the best approach either...

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38881553).

File: torchmultimodal/utils/attention.py (hunk 5 above)
Comment (langong347) [reply torchmultimodal/utils/attention.py:30]: nit: If the more common case is passing without either batch or head dims, perhaps update the docstring to reflect that expectation? Like:

```
Accepting two types of attention masks:
        - Causal: masks that prevent attending to future positions of dimensions
            ``(query_seq_len, key_seq_len)``
        - Padding: masks that prevent attending to token paddings of dimensions
            ``(seq_len)``
```

Comment (ebsmothers) [reply torchmultimodal/utils/attention.py:30]: Oh sorry I should clarify... I mean that the common cases are `(b, query_seq_len, key_seq_len)` or `(b, seq_len)`. So in either case we still have `b`, it is just `h` that is missing

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38881553).

**Comment by ebsmothers** [comment]
@ebsmothers has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38881553).
